### ISSUE TITLE
`around_` inadvertently affects the test context

### ISSUE THREAD

**Issue by sellout** [hspec/hspec#850]
`around_`, and presumably other hooks, behave differently than I would expect – their effects encapsulate `evaluateExample`, not just the test case. And that leads to some confusing behavior.

In the below repro, `CheckEnv` checks whether an env var in the test matches its value outside the test. `withEnv` brackets the env var.

The env var is set to `"success"` for the duration of the suite.

The first test case locally overrides the var to `"failure"` and captures it. The test fails (as it should) because the captured value doesn’t match the value retrieved in `evaluateExample`.

I would expect the second test to behave the same way – `around_ foo` behaving as if each case were written as `foo $ <test case>`. However, the test succeeds because `around_` also wraps the call to `evaluateExample`, not simply the test case code. So both the expected and actual values are `"failure"`.

```haskell
import Control.Applicative (liftA2)
import Control.Exception (bracket)
import Data.IORef (readIORef, newIORef, writeIORef)
import System.Environment (getEnv, setEnv)
import Test.Hspec (around_, describe, hspec, it, runIO)
import Test.Hspec.Core.Spec (
  Example (..),
  FailureReason (..),
  Result (..),
  ResultStatus (..))

envVar :: String
envVar = "TEST_RESULT"

data CheckEnv = CheckEnv String

instance Example (IO CheckEnv) where
  type Arg (IO CheckEnv) = ()
  evaluateExample checkEnv _ action _ = do
    ref <- newIORef (Result "didn’t run" Success)
    action $ \() -> do
      result <- liftA2
        (\expected (CheckEnv actual) ->
          Result "ran"
          (if expected == actual
            then Success
            else Failure Nothing (ExpectedButGot Nothing expected actual)))
        (getEnv envVar)
        checkEnv
      writeIORef ref result
    readIORef ref

withEnv :: (String, String) -> IO a -> IO a
withEnv (var, val) action =
  bracket (getEnv var) (setEnv var) $ \_ -> do
    setEnv var val
    action

main :: IO ()
main = hspec $ do
  runIO $ setEnv envVar "success"

  describe "checkEnv manual" $
    it "should fail" $
      withEnv (envVar, "failure") $ do
        CheckEnv <$> getEnv envVar
  describe "checkEnv around" $ around_ (withEnv (envVar, "failure")) $
    it "should fail" $ do
      CheckEnv <$> getEnv envVar
```

```
checkEnv manual
  should fail [✘]
    ran
checkEnv around
  should fail [✔]
    ran

Failures:

  test/spec/Spec.hs:55:5: 
  1) checkEnv manual should fail
       expected: success
        but got: failure
```

Lest this be seen as too contrived, this is a reduced example from the actual one I enountered, where we use `around_ inTempDirectory` to get each test case to write in its own temp directory, but that caused hspec-golden to look for the golden files (which it does in `evaluateExample`) in the locally-scoped temp directory rather than in the usual place.

**Comment by sol** [comment]
If I change your `Example` instance to

```haskell
instance Example (IO CheckEnv) where
  type Arg (IO CheckEnv) = ()
  evaluateExample example _ hook _ = do
    ref <- newIORef $ CheckEnv "success"
    liftA2
        (\expected (CheckEnv actual) ->
          Result "ran"
          (if expected == actual
            then Success
            else Failure Nothing (ExpectedButGot Nothing expected actual)))
        (getEnv envVar)
        (hook (\() -> example >>= writeIORef ref) >> readIORef ref)
```
then I get the behavior you expect.

Note that I renamed `checkEnv` to `example` and `action` to `hook`.

**TL;DR: What you need to do is apply the `hook` to the `example` and do any other bookkeeping outside of that.**

From what I understand, this is a bug in `hspec-golden`.  Please report it there.  If somebody is going to fix this in `hspec-golden` then I can offer to take a quick look at the fix and offer third-party feedback.

**Comment by sellout** [comment]
@sol – thanks! Opened a ticket there, closing this one.
